src/RequestHandler/inprocess/inprocessapplication.cpp

```diff
@@ -227,58 +238,49 @@ IN_PROCESS_APPLICATION::ShutDownInternal()
         // as nothing can be done
         DeleteFile(m_struLogFilePath.QueryStr());
     }
-
-Finished:
-
-    if (fLocked)
-    {
-        ReleaseSRWLockExclusive(&m_srwLock);
-    }
-    return;
 }
 
+__override
 VOID
 IN_PROCESS_APPLICATION::Recycle(
     VOID
 )
 {
-    BOOL  fLockAcquired = FALSE;
-    // We need to guarantee that recycle is only called once, as calling pHttpServer->RecycleProcess
-    // multiple times can lead to AVs. 
-    if (m_fRecycleCalled)
+    try
     {
-        goto Finished;
-    }
+         // We need to guarantee that recycle is only called once, as calling pHttpServer->RecycleProcess
+        // multiple times can lead to AVs. 
+        if (m_fRecycleCalled)
+        {
+            return;
+        }
 
-    AcquireSRWLockExclusive(&m_srwLock);
-    fLockAcquired = TRUE;
+        {
+            std::unique_lock<std::mutex> lock(m_srwLock);
```

Comment:
this is the part I personally don't prefer. The code depends on the destructor of unique_lock to do unlock  once the object is out of scope. This makes the code no that easy to be read.

Reply:
Yeah, but C++ guarantees that it would run vs a person having to remember to add `Finished` label and undo it there.

Reply:
my understanding is that SWRLOCK is faster than C++ unique_lock as later requires more resource.
Also even performance is not a concern, I would prefer doing unlock specifically for easy code readability and maintenance. The rule is that optimize code for the reader instead of the writer.  For instance, it is not easy to understand why IISExpress scenario is not in else  section for if (!m_pHttpServer->IsCommandLineLaunch())

Reply:
I removed unique_lock replacing it with SWR wrapper. I'll try improve readability.

Reply:
I rewrote the method a bit, `m_pHttpServer->RecycleProcess` call doesn't require lock because there is nothing else calling method on it and first part would prevent double call.
Notice that we never actually set `m_fRecycleCalled` value to true before so the code had a bug.

Reply:
I wan't us to use RAII in cases where resource is being aquired and returned in a scope of one method, it would greatly simplify the logic and allows to use returns instead goto's and labels.